xdsl/dialects/experimental/air.py

```diff
@@ -0,0 +1,810 @@
+
+
+@irdl_op_definition
+class LaunchTerminatorOp(IRDLOperation):
+    name = "air.launch_terminator"
+
+    traits = traits_def(lambda: frozenset([HasParent(LaunchOp)]))
+
+
+@irdl_op_definition
+class LaunchOp(IRDLOperation):
+    name = "air.launch"
+
+    sym_name = attr_def(StringAttr)
+    async_dependencies = var_operand_def(AsyncTokenAttr())
+    sizes = var_operand_def(IndexType())
+    launch_operands = var_operand_def(AnyAttr())
+    async_token = result_def(AsyncTokenAttr)
+
+    traits = frozenset(
+        [IsolatedFromAbove(), SingleBlockImplicitTerminator(LaunchTerminatorOp)]
+    )
+
+    irdl_options = [AttrSizedOperandSegments()]
+
+    def __init__(
+        self,
+        sym_name: StringAttr,
+        async_dependencies: list[Operation | SSAValue],
+        sizes: list[Operation | SSAValue],
+        launch_operands: list[Operation | SSAValue],
+    ):
+        super().__init__(
+            attributes={"sym_name": sym_name},
+            operands=[async_dependencies, sizes, launch_operands],
+            result_types=[AsyncTokenAttr()],
+        )
+
+
+@irdl_op_definition
+class HerdPipelineOp(IRDLOperation):
+    name = "air.pipeline"
+
+    traits = frozenset([HasParent(HerdOp)])
+
+
+@irdl_op_definition
+class PipelineGetOp(IRDLOperation):
+    name = "air.pipeline.get"
+
+    src0 = operand_def(Attribute)
+    src1 = operand_def(Attribute)
+    results = var_result_def(Attribute)
+
+    def __init__(
+        self,
+        src0: Operation | SSAValue,
+        src1: Operation | SSAValue,
+        result_types: list[Attribute],
+    ):
+        super().__init__(operands=[src0, src1], result_types=result_types)
+
+
+@irdl_op_definition
+class PipelinePutOp(IRDLOperation):
+    name = "air.pipeline.put"
+
+    dst0 = operand_def(Attribute)
+    dst1 = operand_def(Attribute)
+    opers = var_operand_def(Attribute)
+
+    def __init__(
+        self,
+        dst0: Operation | SSAValue,
+        dst1: Operation | SSAValue,
+        opers: list[Operation | SSAValue],
+    ):
+        super().__init__(operands=[dst0, dst1, opers])
+
+
+@irdl_op_definition
+class PipelineStageOp(IRDLOperation):
+    name = "air.pipeline.stage"
+
+    opers = var_operand_def(Attribute)
+    results = var_result_def(Attribute)
+
+    traits = frozenset([HasParent(HerdPipelineOp)])
+
+    def __init__(
+        self, opers: list[Operation | SSAValue], result_types: list[Attribute]
+    ):
+        super().__init__(operands=[opers], result_types=result_types)
+
+
+@irdl_op_definition
+class PipelineTerminatorOp(IRDLOperation):
+    name = "air.pipeline.terminator"
+
+    opers = var_operand_def(Attribute)
+
+    traits = frozenset([HasParent(HerdPipelineOp), IsTerminator()])
+
+    def __init__(self, opers: list[Operation | SSAValue]):
+        super().__init__(operands=[opers])
+
+
+@irdl_op_definition
+class PipelineYieldOp(IRDLOperation):
+    name = "air.pipeline.yield"
```

Comment:
If this isnt using an exotic syntax, this could inherit from AbstractYieldOperations

Reply:
`PipelineYieldOp` has a variadic operand. Operands don't seem to be supported by `AbstractYieldOperation`?